pkg/errors/withstatus.go

```diff
@@ -0,0 +1,58 @@
+package errors
+
+import (
+	"errors"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/runtime/protoiface"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+)
+
+// Domain is the domain used for all errors.
+const Domain = "authzed.com"
+
+// WithCodeAndDetails returns a gRPC status message containing the error's message, the given
+// status code and any supplied details.
+func WithCodeAndDetails(err error, code codes.Code, details ...protoiface.MessageV1) *status.Status {
+	status, err := status.New(code, err.Error()).WithDetails(details...)
+	if err != nil {
+		panic(err)
+	}
+	return status
+}
+
+// ForReason returns an ErrorInfo block for a specific error reason as defined in the V1 API.
+func ForReason(reason v1.ErrorReason, metadata map[string]string) *errdetails.ErrorInfo {
+	return &errdetails.ErrorInfo{
+		Reason:   v1.ErrorReason_name[int32(reason)],
+		Domain:   Domain,
+		Metadata: metadata,
+	}
+}
+
+func WithCodeAndReason(err error, code codes.Code, reason v1.ErrorReason) error {
+	metadata := map[string]string{}
+
+	var hasMetadata HasMetadata
+	if ok := errors.As(err, &hasMetadata); ok {
+		metadata = hasMetadata.DetailsMetadata()
+	}
+
+	status, err := status.New(code, err.Error()).WithDetails(ForReason(reason, metadata))
```

Comment:
Why handle the error here instead of calling the function you just defined above?